[CMake] Revamp detection of _Float16 in Linux

dpino opened the pull request:
#### 974b49251d2830763203253516442f4e01917c19
<pre>
[CMake] Revamp detection of _Float16 in Linux
<a href="https://bugs.webkit.org/show_bug.cgi?id=277900">https://bugs.webkit.org/show_bug.cgi?id=277900</a>

Reviewed by Michael Catanzaro and Justin Michaud.

For platforms not COCOA, detect _Float16 support via successful compilation of
small program using _Float16.

* Source/WTF/wtf/PlatformHave.h:
* Source/cmake/WebKitCompilerFlags.cmake:

Canonical link: <a href="https://commits.webkit.org/282131@main">https://commits.webkit.org/282131@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/269dfafd752e3aa2338fb686f933693d4e60ee9d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62141 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41495 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14733 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66121 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12686 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49181 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13026 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50081 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8781 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65210 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38540 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53822 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30885 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35212 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11084 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11617 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/55237 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56953 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11389 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67850 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/61383 "Found 1 new JSC stress test failure: stress/scoped-arguments-table-should-be-tolerant-for-oom.js.bytecode-cache (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6084 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11154 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57457 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6110 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53795 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57686 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5021 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83147 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9356 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37295 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14575 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38379 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39475 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38124 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -278,7 +278,7 @@
 #endif
 
-#if ((CPU(ARM64) || (CPU(X86_64) && !PLATFORM(PLAYSTATION))) && defined(__FLT16_MANT_DIG__))
+#if !OS(LINUX) && !PLATFORM(PLAYSTATION) && (CPU(ARM64) || CPU(X86_64))
```
I would recommend doing the opposite: use the compiler test everywhere except Cocoa platforms.

Notably, whether `_Float16` is available is really complicated. Look at https://github.com/simd-everywhere/simde/pull/1182 and notice that it depends not just on architecture, but also on compiler, compiler version, and programming language. (E.g. on aarch64 it's only available in C++ since GCC 13, or in C since GCC 7, or in Clang since Clang 15). So this condition doesn't come close to detecting it properly. But I'm not sure whether it actually matters whether we are Linux or PlayStation or FreeBSD or Fuschia....

dpino replied:
Thank you Michael, very interesting. I reworked the patch according to your comment.

mcatanzaro commented on Source/cmake/WebKitCompilerFlags.cmake:
```diff
@@ -460,6 +460,19 @@ int main() {
     unset(CMAKE_REQUIRED_FLAGS)
 endif ()
 
+if (WTF_OS_LINUX)
```
Surely you can run this test on every OS? There's nothing Linux-specific about it.

dpino replied:
I realised COCOA is not an OS, it's a PLATFORM, so I reworked the patch. To make the patch symmetric and guarded the CMake check with `NOT WTF_PLATFORM_COCOA`.

mcatanzaro replied:
Looks good.

justinmichaud reviewed (approved):
r=me

dpino committed:
[CMake] Revamp detection of _Float16 in Linux
https://bugs.webkit.org/show_bug.cgi?id=277900

Reviewed by Michael Catanzaro and Justin Michaud.

For platforms not COCOA, detect _Float16 support via successful compilation of
small program using _Float16.

* Source/WTF/wtf/PlatformHave.h:
* Source/cmake/WebKitCompilerFlags.cmake:

Canonical link: https://commits.webkit.org/282131@main
